File: kalman_python_2d.py
# -*- coding: utf-8 -*-
"""
@author: gerardo
"""

## Creado por Juan Gerardo Castrejon
## Kalman extendido 
## 06/09/2019
#%%
import numpy as np
from numpy.random import randn
from numpy.linalg import inv
import matplotlib.pyplot as plt

#%%
F = np.array([[1, 1],[0, 1]])
H = np.array([1, 0])
G = np.array([0.5, 1])
g = 1

Q = np.array([[0, 0],[0, 0]])
R = 1

#P0 = np.array([])
x0 = np.array([100, 0])

n = 6
u = -g

x = np.zeros((2,n))
z = np.zeros((n,1))
w = 0.5*randn(n,1)

#%%
def xs(xa): 
    return F@xa + G*u

#%%
x[:,0] = x0
for k in range(1,n):
    x[:,k] = xs(x[:,k-1])
    z[k] = H@x[:,k] + w[k]

#%%
xp = np.zeros((2,n))
xc = np.zeros((2,n))
Pp = np.zeros((2,2,n))
Pc = np.zeros((2,2,n))

xc0 = np.array([95, 1])
Pc0 = np.array([[10, 0], [0, 1]])

#%%
xc[:,0] = xc0
Pc[:,:,0] = Pc0

for k in range(1,n):
    # Prediccion
    xp[:,k] = F@xc[:,k-1] + G*u
    Pp[:,:,k] = F@Pc[:,:,k-1]@F.T + Q
    
    # Correccion
    S = H@Pp[:,:,k]@H.T + R
    K = Pp[:,:,k]@H.T/S #OBSERVE: en python inv no puede aplicarse a escalares
    xc[:,k] = xp[:,k] + K*(z[k] - H@xp[:,k])
    Pc[:,:,k] = Pp[:,:,k] - K*S@K.T #OBSERVE: en python @ no puede aplicarse a escalares


#%%
t = range(0,6)
plt.plot(t, x[0,:], t, z, t, xc[0,:])
plt.axis([1,6,85,101])
plt.grid(True)
plt.legend(['True', 'Measurement','Estimate'])
plt.show()

File: kalman_python_2d.jl
using LinearAlgebra
using PyPlot
pygui(true)


#%%
F = [1 1; 0 1]
H = [1 0]
G = [0.5; 1]
g = 1

Q = [0 0;0 0]
R = 1

x0 = [100; 0]

n = 6
u = -g

x = zeros(2,n)
z = zeros(n,1)
w = 0.5*randn(n,1)

#%%
xs(xa) = F*xa + G*u'

#%%
x[:,1] = x0
for k = 2:n
    x[:,k] = xs(x[:,k-1])
    z[k] = H⋅x[:,k] + w[k] #OBSERVE: z recibe un escalar por lo que se usa dot() y no *
end

#%%
xp = zeros(2,n)
xc = zeros(2,n)

Pp = zeros(2,2,n)
Pc = zeros(2,2,n)

xc0 = [95; 1]
Pc0 = [10 0;0 1]

#%%
xc[:,1] = xc0
Pc[:,:,1] = Pc0

for k = 2:n
    # Prediccion
    xp[:,k]   = F*xc[:,k-1] + G*u
    Pp[:,:,k] = F*Pc[:,:,k-1]*F' + Q

    # Correccion
    S = H*Pp[:,:,k]⋅H' + R #OBSERVE: dot() en vez de *
    K = Pp[:,:,k]*H'*inv(S)
    xc[:,k] = xp[:,k] + K*(z[k] - H⋅xp[:,k]) #OBSERVE: dot() en vez de *
    Pc[:,:,k] = Pp[:,:,k] - K*S*K'
end

#%%
figure
t = 1:6
plot(t, x[1,:], t, z, t, xc[1,:])
axis([1,6,85,101])
grid(true)
legend(["True", "Measurement", "Estimate"])
